ocaml-lsp-server/src/diagnostics.ml

```diff
@@ -0,0 +1,64 @@
+open Import
+
+module Uri = struct
+  include Uri
+
+  let compare x y = Ordering.of_int (Uri.compare x y)
+end
+
+module Uri_c = Comparable.Make (Uri)
+module Uri_set = Uri_c.Set
+
+type t =
+  { dune : (int, Uri.t * Diagnostic.t) Table.t
+  ; merlin : (Uri.t, Diagnostic.t list) Table.t
+  ; send : PublishDiagnosticsParams.t -> unit Fiber.t
```

Comment:
Why is this dynamic? I would expect the method of sending diagnostics to be general to all the diagnostics

Reply:
It's needed to avoid circular dependencies. I'll see if I can do it in another way though.

Reply:
I tried to, but I really don't see a good alternative. It's possible to define State.t and Dune.t in a mutually recursive manner, but that means leaking the internal fields of Dune.t. That seems much uglier than passing an ugly send callback.

Reply:
Alright! I'm not familiar enough with the whole architecture to have suggestions here, so I trust your judgement 🙂

Reply:
I don't necessarily trust my own judgement, but it's just that I'm not sure what to do to improve this code without introducing other issues. So an improvement is definitely is possible, but it will require additional thinking and unlikely to come in this PR unless some concrete suggestions are made.